Added a growth progress bar to CityButton.

drwhut opened the pull request:
Hey!

I was having a look at issue #663 and saw that there wasn't a "growth bar" for the player's cities on the world screen. So I created some functions in `CityInfo` to help determine if the city is growing, starving, and how many turns it would take to grow or starve - I then used these functions not only in the city details screen but on the `CityButton` with a progress bar.

I hope the changes are alright! Let me know if you're not keen on something I've done.

yairm210 commented:
Code looks ok, can you post a picture of how it looks?

r3versi commented on core/src/com/unciv/ui/tilegroups/CityButton.kt:
```diff
@@ -155,6 +159,52 @@ class CityButton(val city: CityInfo, internal val tileGroup: WorldTileGroup, ski
+                turnLabel.color = Color.RED
+            } else {
+                turnLabel = "-".toLabel()
+                turnLabel.color = Color.WHITE
```
I think there should some upper bound on turns to growth and starvation. Probably if it's higher than 99 you can show an ∞ symbol? 
Also in case the population is stable maybe show ∞? (or don't show at all the growth group?)

In case of starvation why is the number of turns shown as negative?

drwhut replied:
I have tried to display the infinity symbol, but for some reason it doesn't show it in-game. Maybe the graphics library isn't configured for unicode properly?

r3versi replied:
To display a unicode symbol you have to add first to the set of symbols loaded: https://github.com/yairm210/Unciv/blob/c6b96f4f5b711aab3a77a249767b522775e312bd/core/src/com/unciv/ui/utils/Fonts.kt#L21

r3versi commented on core/src/com/unciv/ui/tilegroups/CityButton.kt:
```diff
@@ -155,6 +159,52 @@ class CityButton(val city: CityInfo, internal val tileGroup: WorldTileGroup, ski
+            var growthPercentage = city.population.foodStored / city.population.getFoodToNextPopulation().toFloat()
+            if (growthPercentage < 0) growthPercentage = 0.0f
+
+            val growthBar = ImageGetter.getProgressBarVertical(2f, groupHeight, growthPercentage,
```
In case of starvation how about showing a full RED bar? This way it could be easier to spot which cities have negative growth.

r3versi replied:
As for the growth bar, maybe use the secondary color of the civ (or a less _flashy_ green?) when the city is growing?
I think that with some civs colors the light green could look weird:
![off3](https://user-images.githubusercontent.com/17114100/71821452-7dabf000-3092-11ea-8c12-85f6b47c5304.png)

r3versi commented on core/src/com/unciv/ui/tilegroups/CityButton.kt:
```diff
@@ -155,6 +159,52 @@ class CityButton(val city: CityInfo, internal val tileGroup: WorldTileGroup, ski
+    private fun getPopulationGroup(showGrowth: Boolean): Group {
+        val group = Group()
+
+        val populationLabel = city.population.population.toString().toLabel()
```
populationLabel should be colored using the secondary color of the civ owning the city:
![Off2](https://user-images.githubusercontent.com/17114100/71821329-23ab2a80-3092-11ea-82a4-136e8c3b1781.png)

r3versi commented on core/src/com/unciv/logic/city/CityInfo.kt:
```diff
@@ -192,6 +194,35 @@ class CityInfo {
+    }
+
+    /** Take -1 to mean infinity. */
+    fun getNumTurnsToNewPopulation(): Int {
```
Maybe instead of `-1` you can return `null` in case the city is not growing? Might be more expressive, not sure though

drwhut replied:
I did try to return `null`, but Android Studio was throwing errors about returning null in a non-null type, so I went with `-1` instead.

r3versi replied:
Yeah, if you return `null` the type returned by the functions should be `Int?` to reflect the nullability of the returned type.

drwhut commented:
Here's the list of changes I've made:

* Changed the population label's colour to be the secondary colour of the civ.
* If a city is growing, the growth bar is a duller green than before.
* If a city is starving, the growth bar is a full red, and the number is no longer negative.
* If the number of turns next to the growth bar is more than 99, it shows a dash.

Here's some pictures as well:
![unciv-growth](https://user-images.githubusercontent.com/13645454/71828525-653fd380-309a-11ea-9147-2c9365ff0074.png)
![unciv-nogrowth](https://user-images.githubusercontent.com/13645454/71828527-653fd380-309a-11ea-8fa4-88790737f5b3.png)
![unciv-population](https://user-images.githubusercontent.com/13645454/71828528-65d86a00-309a-11ea-9279-273a2f87ae9d.png)
![unciv-starving](https://user-images.githubusercontent.com/13645454/71828529-65d86a00-309a-11ea-9dff-956a3bc5fadd.png)

r3versi commented:
The population `Group` isn't properly sized and overflows into the city name:
![Off1](https://user-images.githubusercontent.com/17114100/71830077-3b8ca980-30a6-11ea-907b-d5ed2bb6fbfa.png)

r3versi commented on core/src/com/unciv/ui/tilegroups/CityButton.kt:
```diff
@@ -160,9 +160,12 @@ class CityButton(val city: CityInfo, internal val tileGroup: WorldTileGroup, ski
     }
 
     private fun getPopulationGroup(showGrowth: Boolean): Group {
+        val growthGreen = Color(0.0f, 0.5f, 0.0f, 1.0f)
```
This is much nicer, but there are conflicts (imo) with several civilizations:
![BadColor](https://user-images.githubusercontent.com/17114100/71830679-88bd4b00-30a7-11ea-828c-a0d655d6bfd0.png)
![BadColor2](https://user-images.githubusercontent.com/17114100/71830680-8955e180-30a7-11ea-8bba-b41a41973a5b.png)
![BadColor3](https://user-images.githubusercontent.com/17114100/71830681-8955e180-30a7-11ea-8039-40432a513858.png)

Maybe you can use the secondary color in case the city is growing?

drwhut replied:
Fair enough, I was trying to replicate Civ 5, but yeah I agree with those civ colours it's a bit bad.

Do you want both the growth bar *and* the number to be the secondary colour, or just the number or just the growth bar?

r3versi replied:
Don't know tbh. Maybe you can try and see how it looks? Or just wait for @yairm210 opinion

drwhut replied:
Yeah, I'll see what @yairm210 says.

In my opinion, I think at least one of the bar or number (probably the bar) should stay the green colour, only because it may be initially confusing for new players if they see 2 numbers and a bar all the same colour, as they may not understand what differentiates them. Whereas if say the bar is green, it's the same colour as food, so the player *may* associate it with food at least.

yairm210 commented:
I think the bar should remain here/red because it will always contrast with the black part of the bar
The number will not always contrast, so if should probably be the secondary colour
This will mean that for the new player, the number colour will be the same whether it's for growth or starvation, but the colour of the bar should be enough to alert then that is not the same

drwhut commented:
I agree. Here's what it looks like now:

![unciv-newnumber](https://user-images.githubusercontent.com/13645454/71840236-caa0be00-30b4-11ea-8f10-e66c5bd46934.png)

* Instead of dashes, it is now an infinity symbol.
* The number of turns to growth/starvation functions return `null` instead of `-1` to represent infinity.
* Adjusted the size of the population group to account for large numbers.
* Changed the colour of the growth number to match the secondary colour of the civ.

r3versi commented:
@drwhut don't know why, but after a city is conquered (the moment after you click on annex/raze/puppet) the bar goes wild.
![Capture](https://user-images.githubusercontent.com/17114100/71994857-c520b080-3239-11ea-9bff-e241ab555f43.png)

However on next turn it gets back to normality.

drwhut commented:
Weird... sorry that I didn't test that case. Is it possible for you to send me a save file so I can debug it?

r3versi commented:
Sure! [SaveFile](https://pastebin.com/raw/gEf5Mnwr)
It is the city named `DebugMe`, bottom right of the map.

drwhut commented:
The game can't load the file @r3versi - I've just had a look at the contents, and my autosaves are in a JSON format, but that file isn't. Is it encrypted, or just borked?

r3versi commented:
The save file is compressed with gzip and encoded in Base64. To load into the game you can copy the text > Menu > Load Game > Copy from clipboard. 

Otherwise just run this to decode and decompress the text file i sent you:

    base64 -d data.txt | gzip -d > SaveFile.json

drwhut commented:
Fixed! #1646
